`ConcurrentModificationException` in `HealthCheckedEndpointGroup.<init>`

```
Caused by: java.util.ConcurrentModificationException: null
	at java.base/java.util.HashMap$Values.forEach(HashMap.java:979)
	at com.linecorp.armeria.client.endpoint.healthcheck.HealthCheckedEndpointGroup.<init>(HealthCheckedEndpointGroup.java:126)
	at com.linecorp.armeria.client.endpoint.healthcheck.AbstractHealthCheckedEndpointGroupBuilder.build(AbstractHealthCheckedEndpointGroupBuilder.java:147)
```

--- a/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java
+++ b/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java
@@ -56,7 +56,7 @@ public class DynamicEndpointGroup extends AbstractListenable<List<Endpoint>> imp
         final List<Endpoint> newEndpoints;
         endpointsLock.lock();
         try {
-            List<Endpoint> newEndpointsUnsorted = Lists.newArrayList(endpoints);
+            final List<Endpoint> newEndpointsUnsorted = Lists.newArrayList(endpoints);
             newEndpointsUnsorted.add(e);
             endpoints = newEndpoints = ImmutableList.sortedCopyOf(newEndpointsUnsorted);
         } finally {

--- a/core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroup.java
+++ b/core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroup.java
@@ -42,6 +42,7 @@ import org.slf4j.LoggerFactory;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.MoreObjects;
 import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
 import com.spotify.futures.CompletableFutures;
 
 import com.linecorp.armeria.client.ClientFactory;
@@ -123,7 +124,11 @@ public final class HealthCheckedEndpointGroup extends DynamicEndpointGroup {
         updateCandidates(delegate.initialEndpointsFuture().join());
 
         // Wait until the initial health of all endpoints are determined.
-        contexts.values().forEach(ctx -> ctx.initialCheckFuture.join());
+        final Map<Endpoint, DefaultHealthCheckerContext> copy;
+        synchronized (contexts) {
+            copy = ImmutableMap.copyOf(contexts);
+        }
+        copy.values().forEach(ctx -> ctx.initialCheckFuture.join());
     }
 
     private void updateCandidates(List<Endpoint> candidates) {

--- a/core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroup.java
+++ b/core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroup.java
@@ -42,7 +42,6 @@ import org.slf4j.LoggerFactory;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.MoreObjects;
 import com.google.common.collect.ImmutableList;
-import com.google.common.collect.ImmutableMap;
 import com.spotify.futures.CompletableFutures;
 
 import com.linecorp.armeria.client.ClientFactory;
@@ -124,11 +123,11 @@ public final class HealthCheckedEndpointGroup extends DynamicEndpointGroup {
         updateCandidates(delegate.initialEndpointsFuture().join());
 
         // Wait until the initial health of all endpoints are determined.
-        final Map<Endpoint, DefaultHealthCheckerContext> copy;
+        final List<DefaultHealthCheckerContext> snapshot;
         synchronized (contexts) {
-            copy = ImmutableMap.copyOf(contexts);
+            snapshot = ImmutableList.copyOf(contexts.values());
         }
-        copy.values().forEach(ctx -> ctx.initialCheckFuture.join());
+        snapshot.forEach(ctx -> ctx.initialCheckFuture.join());
     }
 
     private void updateCandidates(List<Endpoint> candidates) {

